### PR TITLE
Implement RequestServiceBuilder

### PR DESCRIPTION
**Pull request by seanpianka**
This is based on work by @aetf [in this commit](https://github.com/Aetf/routerify/commit/6627e965dea0c2bc07d50975c445b687dff3b1a9), which I found while reviewing the public forks of Routerify. 

> "Completely decouple the route processing from
> request handling, enabling using the library in
> AWS lambda rust runtime."

Running within AWS Lambda is my [usecase](https://github.com/seanpianka/rust-routerify-aws-lambda-quickstart)/focus of study. The original author suggested avoiding the steps of starting a Hyper server and client on every request (discussion is on the original commit webpage). I found this change useful, so we should consider adding this separation for other users who may require this as a part of their workflow.

The change has been written using the updated Tokio 1.0/Hyper 0.14 fork referenced in #35 and so is blocked waiting for its merge.

### CODE REVIEW

**Comment by Aetf** [comment]
I don't think `hyper::accept::poll_fn` and `futures::future::poll_fn` are the same thing, and the latter should be used.

**Comment by seanpianka** [comment]
I've updated the test to use `futures::future::poll_fn` and it still succeeds. However, I am now seeing the following warning:
```rust
warning: unused `futures::future::PollFn` that must be used
   --> src/service/request_service.rs:131:9
    |
131 |         poll_fn(|_| -> Poll<Result<(), Error>> { Poll::Ready(Ok(())) });
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `#[warn(unused_must_use)]` on by default
    = note: futures do nothing unless you `.await` or poll them
```

You mentioned in our original discussion that invoking `poll_fn` is not necessary in this specific case of this unit test... I assume that's because nothing is using the return value from `poll_fn` here?

**Comment by Aetf** [comment]
The return value of `poll_fn` is `impl Future<Output=Result<()>>`, so you need to await on it and check for error. See https://github.com/Aetf/fit2/blob/main/src/lib.rs#L68

**Comment by seanpianka** [comment]
Aha! That clarifies things, thanks for the example. I have updated the test to call `poll_ready` of the `RequestService`, then await `poll_fn`'s returned future, and then panic if that future's result is `Err(e)`.

The test still passed and the warning has disappeared, so things are looking good!

File: src/service/request_service.rs
Comment (Aetf) [inline src/service/request_service.rs:133]: You can simply do `.await.expect("request service is not ready")`

Comment (seanpianka) [reply src/service/request_service.rs]: Fixed!

**Comment by rousan** [comment]
Send PR to develop

**Comment by rousan** [comment]
@seanpianka , The PR seems to have conflict, we should resolve it before merging.

**Comment by seanpianka** [comment]
Hi @rousan, I've rebased from the current develop and have fixed the merge conflicts that were caused after the immutability updates. All the tests have passed, so I'm just waiting on approvals before merging this into develop.

File: src/service/request_service.rs
Comment (gsserge) [inline src/service/request_service.rs:95]: There seem to be no reason for a separate impl for `build()`; it can be moved over in the impl above with `new()`.

**Review by gsserge — APPROVED** [review]
Looks good! It would be awesome to have a full example of Routerify with AWS Lambda in the `examples/` folder.

**Comment by seanpianka** [comment]
I've got a write up that describes the setup, but I suppose I could add the example to this repository as well.

**Comment by seanpianka** [comment]
Okay, I've added an example (and tests for the example) of using Routerify for an AWS Lambda function. It seems that running `$ cargo test --features="all"` does not invoke these tests. I am able to explicitly invoke them manually with `cargo test --example aws_lambda tests`, shall I add this to the GitHub actions workflow?

I'd also appreciate a quick review of the example 🙂 I'm not sure if my implementation is the most efficient, even if it is functional.

File: src/service/request_service.rs
Comment (seanpianka) [reply src/service/request_service.rs:93]: Updated to one impl

**Comment by gsserge** [comment]
In order to invoke tests in the examples the `--all-targets` flag is needed (by default `cargo test` only builds the examples)

```
$ cargo test --verbose --features="all" --all-targets
```

You could add `--all-targets` to `run: cargo test --verbose --features="all"` in `.github/workflows/test.yml`.

Once this PR is merged I think we need to create another and hopefully the last beta, and then finally 2.0.

**Comment by seanpianka** [comment]
I've updated the CI test invocation to include `--all-targets`, so I'll merge this in now.

**Pull request merged**
